stdlib/public/Concurrency/AsyncFailableDropWhileSequence.swift

```diff
@@ -0,0 +1,84 @@
+////===----------------------------------------------------------------------===//
+////
+//// This source file is part of the Swift.org open source project
+////
+//// Copyright (c) 2021 Apple Inc. and the Swift project authors
+//// Licensed under Apache License v2.0 with Runtime Library Exception
+////
+//// See https://swift.org/LICENSE.txt for license information
+//// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+////
+////===----------------------------------------------------------------------===//
+
+import Swift
+
+extension AsyncSequence {
+  @inlinable
+  public __consuming func drop(
+    while predicate: @escaping (Element) async throws -> Bool
+  ) -> AsyncFailableDropWhileSequence<Self> {
+    AsyncFailableDropWhileSequence(self, predicate: predicate)
+  }
+}
+
+@frozen
+public struct AsyncFailableDropWhileSequence<Upstream: AsyncSequence> {
+  public let upstream: Upstream
+  public let predicate: (Upstream.Element) async throws -> Bool
+
+  @inlinable
+  public init(
+    _ upstream: Upstream, 
+    predicate: @escaping (Upstream.Element) async throws -> Bool
+  ) {
+    self.upstream = upstream
+    self.predicate = predicate
+  }
+}
+
+extension AsyncFailableDropWhileSequence: AsyncSequence {
+  public typealias Element = Upstream.Element
+  public typealias AsyncIterator = Iterator
+
+  @frozen
+  public struct Iterator: AsyncIteratorProtocol {
+    @usableFromInline
+    var upstreamIterator: Upstream.AsyncIterator
+
+    @usableFromInline
+    var predicate: ((Upstream.Element) async throws -> Bool)?
+
+    @usableFromInline
+    init(
+      _ upstreamIterator: Upstream.AsyncIterator, 
+      predicate: @escaping (Upstream.Element) async throws -> Bool
+    ) {
+      self.upstreamIterator = upstreamIterator
+      self.predicate = predicate
+    }
+
+    @inlinable
+    public mutating func next() async throws -> Upstream.Element? {
+      while let predicate = self.predicate {
+        guard let element = try await upstreamIterator.next() else {
+          return nil
+        }
+        do {
+          if try await predicate(element) == false {
+            self.predicate = nil
+            return element
+          }
+        } catch {
+          self.predicate = nil
+          throw error
```

Comment:
Double-checking an assumption -- downstream is assumed to release last reference to this iterator, so that upstreamIterator is also released, in turn?

Reply:
That is correct the lifetime is owned by downstream; Since we have no ref to downstream we don't need to break bonds between operators like we do in Combine. The model is:
`Values -> -> ->`
`Ownership <- <- <-`

That is why all the operators are marked with `__consuming`.